pkg/image/image.go

```diff
@@ -98,108 +98,46 @@ func (i *Info) UsedBy() []string {
-	pullPath := filepath.Join(location, "."+rand.GenerateID(64))
-	glog.V(5).Infof("Pulling %s to temporary file %s", ref, pullPath)
-	defer func() {
-		if err != nil {
-			if err := os.Remove(pullPath); err != nil && !os.IsNotExist(err) {
-				glog.Errorf("Could not remove %s: %v", pullPath, err)
-			}
-		}
-	}()
-
-	pullURL := strings.TrimPrefix(ref.String(), ref.URI()+"/")
-	switch ref.URI() {
-	case singularity.LibraryDomain:
-		config := &library.Config{
-			BaseURL:   auth.GetServerAddress(),
-			AuthToken: auth.GetRegistryToken(),
-		}
-		client, err := library.NewClient(config)
+func Pull(ctx context.Context, location string, ref *Reference, auth *k8s.AuthConfig) (*Info, error) {
+	if ref.URI() == singularity.LocalFileDomain {
+		info, err := sifInfo(strings.TrimPrefix(ref.tags[0], singularity.LocalFileDomain))
 		if err != nil {
-			return nil, fmt.Errorf("could not create library client: %v", err)
+			return nil, fmt.Errorf("could not fetch local SIF info: %v", err)
 		}
-		w, err := os.Create(pullPath)
-		if err != nil {
-			return nil, fmt.Errorf("could not create file to pull image: %v", err)
-		}
-		parts := strings.Split(pullURL, ":")
-		// don't check index out of range since we add :latest by default when parsing ref
-		err = client.DownloadImage(ctx, w, parts[0], parts[1], nil)
-		_ = w.Close()
-		if err != nil {
-			return nil, fmt.Errorf("could not pull library image: %v", err)
-		}
-	case singularity.DockerDomain:
-		var errMsg bytes.Buffer
-		if auth.GetServerAddress() != "" {
-			pullURL = fmt.Sprintf("%s/%s", auth.GetServerAddress(), pullURL)
-		}
-		remote := fmt.Sprintf("%s://%s", singularity.DockerProtocol, pullURL)
-		buildCmd := exec.CommandContext(ctx, singularity.RuntimeName, "build", "-F", pullPath, remote)
-		buildCmd.Env = []string{
-			fmt.Sprintf("PATH=%s", os.Getenv("PATH")),
-			// assume auth.Auth is not needed b/c k8s decodes it into username and password,
-			// see https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/config.go#L284
-			fmt.Sprintf("%s=%s", singularity.EnvDockerUsername, auth.GetUsername()),
-			fmt.Sprintf("%s=%s", singularity.EnvDockerPassword, auth.GetPassword()),
-		}
-		buildCmd.Stderr = &errMsg
-		buildCmd.Stdout = ioutil.Discard
-		err = buildCmd.Run()
-		if err != nil {
-			err = fmt.Errorf("could not build image: %s", &errMsg)
-		}
-	default:
-		err = fmt.Errorf("unknown image registry: %s", ref.URI())
-	}
-	if err != nil {
-		return nil, fmt.Errorf("could not pull image: %v", err)
+		info.Ref = ref
+		return info, nil
 	}
 
-	pulled, err := os.Open(pullPath)
-	if err != nil {
-		return nil, fmt.Errorf("could not open pulled image: %v", err)
-	}
-
-	fi, err := pulled.Stat()
-	if err != nil {
-		return nil, fmt.Errorf("could not fetch file info: %v", err)
+	pullPath := filepath.Join(location, "."+rand.GenerateID(64))
+	glog.V(5).Infof("Pulling %s to temporary file %s", ref, pullPath)
+	cleanup := func() {
```

Comment:
defer -->. if err != nil --> cleanup ?

Reply:
was trying to get rid of named return values